### DISCUSSION TITLE
Connected components labeling perimeter and circularity thresholds

### DISCUSSION THREAD

**Discussion by devopsix** [ImageMagick/ImageMagick#7257] [Help] [answered]
### ImageMagick version

7.1.1-29

### Operating system, version and so on

Windows 10, Powershell

### Description

I would like to use perimeter and circularity thresholds in connected components labeling, but I am struggling to understand how to work with them.

Given a binary image with a black circle and a black square drawn on white background (shapes.png):

![shapes](https://github.com/ImageMagick/ImageMagick/assets/58369032/2818a66d-3d40-49aa-bfdd-4d650a1950ab)

Image size: 200px x 100px
Square edge length 50px
Circle diameter: 50px

```
magick shapes.png \
-define connected-components:verbose=true \
-define connected-components:area-threshold=0 \
-define connected-components:perimeter-threshold=0 \
-connected-components 8 \
-contrast-stretch 0 \
components1.png
Objects (id: bounding-box centroid area mean-color perimeter):
  0: 200x100+0+0 97.6,49.5 15584 gray(255) 364
  2: 50x50+125+25 149.5,49.5 2500 gray(0) 0
  1: 50x50+25+25 49.5,49.5 1916 gray(0) 202
```

![components1](https://github.com/ImageMagick/ImageMagick/assets/58369032/c7c453e2-d3e7-44c7-9c1b-efbac71fe3ee)

- Why does the background (ID 0) have a perimeter of significantly less than 600?
- Why does the square (ID 2) have a perimeter of 0?

`-define connected-components:circularity-threshold=0` added:
```
magick shapes.png \
-define connected-components:verbose=true \
-define connected-components:area-threshold=0 \
-define connected-components:perimeter-threshold=0 \
-define connected-components:circularity-threshold=0 \
-connected-components 8 \
-contrast-stretch 0 \
components2.png
Objects (id: bounding-box centroid area mean-color perimeter circularity):
  0: 200x100+0+0 97.6,49.5 15584 gray(255) 436 1.47804
  2: 50x50+125+25 149.5,49.5 2500 gray(0) 0 0
  1: 50x50+25+25 49.5,49.5 1916 gray(0) 202 0.884377
```

![components2](https://github.com/ImageMagick/ImageMagick/assets/58369032/3b63ae3b-dfdd-4d77-9e4e-df024820d014)

- Why does the background's (ID 0) perimeter change?

Resizing the image to 50%:
```
magick shapes.png \
-resize "50%" -threshold "50%" \
-define connected-components:verbose=true \
-define connected-components:area-threshold=0 \
-define connected-components:perimeter-threshold=0 \
-define connected-components:circularity-threshold=0 \
-connected-components 8 \
-contrast-stretch 0 \
components3.png
Objects (id: bounding-box centroid area mean-color perimeter circularity):
  0: 100x50+0+0 45.6,24.5 4328 gray(255) 1 48204.6
  2: 26x26+62+12 74.5,24.5 672 gray(0) 0 0
```

![components3](https://github.com/ImageMagick/ImageMagick/assets/58369032/a3e81375-7557-458c-abbc-8034ba962304)

- Why does the background (ID 0) have a perimeter of 1 instead of ~300?

Please note that  the circle (ID 1) is merged into the background.

Resizing the image to 50%, `-define connected-components:circularity-threshold=0` removed:
```
magick shapes.png \
-resize "50%" -threshold "50%" \
-define connected-components:verbose=true \
-define connected-components:area-threshold=0 \
-define connected-components:perimeter-threshold=0 \
-connected-components 8 \
-contrast-stretch 0 \
components4.png
Objects (id: bounding-box centroid area mean-color perimeter):
  0: 100x50+0+0 48.3,24.5 3836 gray(255) 1
  2: 26x26+62+12 74.5,24.5 672 gray(0) 0
  1: 26x26+12+12 24.5,24.5 492 gray(0) 1
```

![components4](https://github.com/ImageMagick/ImageMagick/assets/58369032/57b9580f-1faa-412c-962a-9bccc044620c)

- Why would `-define connected-components:circularity-threshold=0` cause the circle (ID 1) getting merged into the background (see above)? It has a non-zero area and a non-zero perimeter and hence should have a non-zero circularity, doesn't it?

**Comment by fmw42** [comment]
threshold of 0 means take everything.  Set the threshold to a higher number and see if anything less that value has been removed from the list and anything above that value is shown in the list.  Try a range of values, anything within the range is kept in the list.

**Reply by devopsix** [reply]
I appreciate your prompt response, @fmw42!

I understand how to use a single threshold value and a range of values. I use thresholds of 0 for finding all components and including perimeter and circularity values in the verbose output. I do this for getting an understanding of the values.

But I cannot make sense of components' perimeter and circularity values.

Take the third example from above:

```
magick shapes.png \
-resize "50%" -threshold "50%" \
-define connected-components:verbose=true \
-define connected-components:area-threshold=0 \
-define connected-components:perimeter-threshold=0 \
-define connected-components:circularity-threshold=0 \
-connected-components 8 \
-contrast-stretch 0 \
components3.png
Objects (id: bounding-box centroid area mean-color perimeter circularity):
  0: 100x50+0+0 45.6,24.5 4328 gray(255) 1 48204.6
  2: 26x26+62+12 74.5,24.5 672 gray(0) 0 0
```` 

- If thresholds of 0 means “take everything”, why is the circle (ID 1) getting merged into the background?
- What does a circularity value greater than 1 mean?
- Why does a square have a lower circularity value than a non-square rectangle?

Finally, I would like to tell apart shapes which can be approximated by circles and squares from shapes which can be approximated by high aspect ratio rectangles. Based on my understanding of the “circularity formula” (`4*pi*area/perimeter^2`) I would expect 1 for perfect circles, ~0.8 for perfect squares and significantly smaller values for high aspect ratio rectangles.

**Comment by fmw42** [comment, marked as answer]
Given

![shapes](https://github.com/ImageMagick/ImageMagick/assets/13458218/618b3e7d-230e-4306-a5c5-8f4ce47c2760)

with

```
Image size: 200px x 100px
Square edge length 50px
Circle diameter: 50px

The expected values for perimeter and circularity are as follows:

The perimeter of a square or rectangle is 2*(width+height)
The perimeter of a circle is pi*diameter
The circularity is 4*pi*area/perimeter^2
The circularity of a circle should be one and other shapes should be smaller.

However, note that for raster shapes the outline will be jagged and so simple shapes are not going to give perfect values, especially when the shapes are small.  The smaller a circle the less circular it will be as the diameter approaches unity.
```

<br>

Taking them one at at time. ---

### First, looking at the perimeter, when I run

```
magick shapes.png -threshold 50% -type bilevel \
-define connected-components:verbose=true \
-define connected-components:area-threshold=0 \
-define connected-components:perimeter-threshold=0 \
-connected-components 8 \
-contrast-stretch 0 \
components1.png
Objects (id: bounding-box centroid area mean-color perimeter):
  0: 200x100+0+0 97.6,49.5 15584 gray(255) 436
  2: 50x50+125+25 149.5,49.5 2500 gray(0) 0
  1: 50x50+25+25 49.5,49.5 1916 gray(0) 202

```
The background (id=0) perimeter should be approximately 2\*(200+100) = 600 so 436 seems wrong. However, the perimeter of the background may exclude the perimeters of the interior shapes (circle and square).  If that is the case, then then  perimeter should be 2\*(200+100) - 2\*(50+50) - pi\*50 = 600 - 200 - 157 = 243, which is way off from 436. 

**The IM developer will need to reply as to how the perimeter is computed for the background**

The square perimeter (id=2) of 0 is not correct as it should be approximately 2\*(50+50) = 200

Likewise the perimeter of the circle (id=1) of 202 is not correct at is should be approximately pi*diameter = pi\*50 = 157

**So looks like a bug in the perimeter computations**


### Next, looking at the circularity, when I run

```
magick shapes.png -threshold 50% -type bilevel \
-define connected-components:verbose=true \
-define connected-components:area-threshold=0 \
-define connected-components:circularity-threshold=0 \
-connected-components 8 \
-contrast-stretch 0 \
components2.png
Objects (id: bounding-box centroid area mean-color circularity):
  0: 200x100+0+0 97.6,49.5 15584 gray(255) 1.47804
  2: 50x50+125+25 149.5,49.5 2500 gray(0) 0
  1: 50x50+25+25 49.5,49.5 1916 gray(0) 0.884377
```

The circularity of the square of area = 50\*50 = 2500 and perimeter of 2\*(50+50) = 200 should be approximately 0.785398 so a value of 0 is obviously wrong

The circularity of the circle of area = pi\*25\*25 = 1963.4954 and perimeter of pi\*50 = 157.0796 should be approximately 1 so the value of 0.884377 is a bit too low, but depends on the exact computation of area and perimeter since we do not have a perfect circle especially when the radius is small in a raster image as it has a jagged perimeter.

**So looks like bugs in the circularity computations**

**So this definitely needs to be looked into by the IM developer.**

**Reply by devopsix** [reply]
Thank you for confirming that something is odd here!

I have filed issue #7259 for this.
